### ISSUE TITLE
got "Internal consistency check failed - please report this bug"

### ISSUE THREAD

**Issue by halturin** [erlang/otp#4774]
Trying to build the project i'm working on and got this error.

```
===> Compiling apps/arweave/src/ar_network.erl failed
ar_network: function request/3+85:
  Internal consistency check failed - please report this bug.
  Instruction: {call_last,3,{f,444},2}
  Error:       {bad_arg_type,
                   {x,0},
                   {t_union,none,none,none,
                       [{{2,{t_atom,[get_block]}},
                         {t_tuple,2,true,#{1 => {t_atom,[get_block]}}}},
                        {{2,{t_atom,[get_block_index]}},
                         {t_tuple,2,true,#{1 => {t_atom,[get_block_index]}}}},
                        {{2,{t_atom,[get_chunk]}},
                         {t_tuple,2,true,#{1 => {t_atom,[get_chunk]}}}},
                        {{2,{t_atom,[get_sync_record]}},
                         {t_tuple,2,true,
                             #{1 => {t_atom,[get_sync_record]},2 => nil}}},
                        {{2,{t_atom,[get_tx]}},
                         {t_tuple,2,true,#{1 => {t_atom,[get_tx]}}}},
                        {{2,{t_atom,[get_wallet_list]}},
                         {t_tuple,2,true,#{1 => {t_atom,[get_wallet_list]}}}},
                        {{2,{t_atom,[get_wallet_list_chunk]}},
                         {t_tuple,2,true,
                             #{1 => {t_atom,[get_wallet_list_chunk]}}}},
                        {{2,{t_atom,[post_block]}},
                         {t_tuple,2,true,#{1 => {t_atom,[post_block]}}}},
                        {{2,{t_atom,[post_tx]}},
                         {t_tuple,2,true,#{1 => {t_atom,[post_tx]}}}}],
                       none},
                   {t_union,none,none,none,
                       [{{2,{t_atom,[get_block]}},
                         {t_tuple,2,true,#{1 => {t_atom,[get_block]}}}},
                        {{2,{t_atom,[get_block_index]}},
                         {t_tuple,2,true,#{1 => {t_atom,[get_block_index]}}}},
                        {{2,{t_atom,[get_chunk]}},
                         {t_tuple,2,true,#{1 => {t_atom,[get_chunk]}}}},
                        {{2,{t_atom,[get_tx]}},
                         {t_tuple,2,true,#{1 => {t_atom,[get_tx]}}}},
                        {{2,{t_atom,[get_wallet_list]}},
                         {t_tuple,2,true,#{1 => {t_atom,[get_wallet_list]}}}},
                        {{2,{t_atom,[get_wallet_list_chunk]}},
                         {t_tuple,2,true,
                             #{1 => {t_atom,[get_wallet_list_chunk]}}}},
                        {{2,{t_atom,[post_block]}},
                         {t_tuple,2,true,#{1 => {t_atom,[post_block]}}}},
                        {{2,{t_atom,[post_tx]}},
                         {t_tuple,2,true,#{1 => {t_atom,[post_tx]}}}}],
                       none}}:

```

here is the source https://github.com/halturin/arweave/blob/d2fbd2936fdeba53923e8001327ac0b29022d4e1/apps/arweave/src/ar_network.erl

**To Reproduce**
clone this repo 
```git clone --recursive https://github.com/halturin/arweave```
check out to the `d2fbd2936fdeba53923e8001327ac0b29022d4e1`
try to build
```
cd arweave
./rebar3 as prod tar
```

**Expected behavior**
compiled project

**Affected versions**
23.3.1

**Comment by halturin** [comment]
here is the problem line https://github.com/halturin/arweave/blob/d2fbd2936fdeba53923e8001327ac0b29022d4e1/apps/arweave/src/ar_network.erl#L186

```request({broadcast, #{}, ?PEERS_JOINED_MAX}, {get_sync_record, []}, 60000).```

If i replacing it with a simple atom "ok" no errors from the compiler.

**Comment by halturin** [comment]
just checked out with 22.3 - works well. something changed on the way to 23.3.1

**Comment by jhogberg** [comment]
Thanks for your report! I've fixed it in https://github.com/erlang/otp/pull/4776, and will release a patch shortly.

**Comment by halturin** [comment]
Thanks a lot! I close it then.

**Comment by jhogberg** [comment]
I ended up having to revert my fix as it caused much bigger problems elsewhere. I'll try to fix it more thoroughly next week. Sorry about that :-(

**Comment by halturin** [comment]
ok, lets reopen it then.

**Comment by halturin** [comment]
is there any idea how it can be workarounded and what exact line in my code caused this issue?

**Comment by jhogberg** [comment]
> is there any idea how it can be workarounded

Add the following near the top of your module:

```
-compile([no_type_opt]).
```

or

```
-compile([no_module_opt]).
```

That will disable the optimizations that the validation pass (erroneously) believes are unsafe in this case.

>  and what exact line in my code caused this issue?

It doesn't have anything to do with specific lines, and it's quite difficult to explain why this happened unless you're already familiar with compiler internals.

Needless to say there's nothing wrong with your code. This is purely a bug in the compiler's validation pass.

**Comment by jhogberg** [comment]
We've released fixes in `OTP-24.0.2` and `OTP 23.3.4.2`, hopefully they won't cause as many problems this time 😄
